### PR TITLE
Add Posting Details Page UI and Query

### LINKED ISSUE #81: Posting Details Page
### Summary and Background
Volunteers can view more details for a particular posting.

### User Story
As an Volunteer, I want to view more details for a posting, in order to make a more informed decision on whether I want to sign up for this posting or not.

### Solution (Core Elements)
From the 'Browse Volunteer Postings' page, the user can click the 'View details' button to bring up more details on the specific posting.
The details include the following:

- role title / event name
- role description (purpose of the volunteer, outline of responsibilities, etc.)
- branch
- date/time
- relevant skills
- point of contact
- deadline to submit

If the volunteer wants to sign up for that posting, they can select the 'Submit availability' button.

### Mockups

For the entire page, go to https://www.figma.com/file/KWaJRNJcOiW8ThZJ8Sl6On/Volunteer-Experience?node-id=157%3A3123.

## Technical
* Create a new component called `VolunteerPostingDetails.tsx` in frontend/src/components/volunteer/posting
* Fetch posting details by using the `posting` GraphQL query with a posting ID passed into the component as a prop ("browse postings list" will have the IDs - see #83)

### PR DESCRIPTION
**Pull request by albertlai431**
## Ticket link
Closes #81

## Implementation description
* Added types for Posting, Branch, Employee, Shift, and Skill
* Created routing for Posting Details Page
* Used PocCard UI from #136 
* Created query to fetch Posting data and the associated Posting Details Page UI

## Steps to test
1. `docker-compose up --build`
2. Navigate to http://localhost:5000/graphql and run the following mutations:

```graphql
# Create Branch
mutation{
  createBranch(branch: {
    name: "NYC"
  }){
    id
    name
  }
}

# Create Skill
mutation{
  createSkill(skill:{
    name: "Skill1"
  }){
    name
  }
}

# Create Posting
mutation {
  createPosting(posting: {
    branchId: 1
    skills: [1]
    employees: []
    title: "Test Posting"
    type: INDIVIDUAL
    status: DRAFT
    description: "lol",
    startDate: "2022-01-20"
    endDate: "2022-01-21"
    autoClosingDate: "2022-01-23"
    numVolunteers: 10
  }) {
    title
    type
    status
    description
    startDate
    endDate
    autoClosingDate
    numVolunteers
  }
}
```
3. Navigate to http://localhost:3000/volunteer/posting/1 and verify that the UI matches that specified in #81

## What should reviewers focus on?
* Verify that the query works and the UI matches that specified in #81

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] I have run the appropriate linter(s)
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  Divider,
+  Button,
+} from "@chakra-ui/react";
+
+import { gql, useQuery } from "@apollo/client";
+import { useParams } from "react-router-dom";
+import { CalendarIcon, TimeIcon, ChevronLeftIcon } from "@chakra-ui/icons";
+import { PostingResponseDTO } from "../../../../types/api/PostingTypes";
+import PocCard from "../../../common/PocCard";
+import colors from "../../../../theme/colors";
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:18]: You shouldn't need to import colors. You can just reference the colors from our theme using a string. See comments below

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+    fetchPolicy: "cache-and-network",
+    onCompleted: (data) => {
+      setPostingDetails(data.posting);
+    },
+    onError: (error) => {
+      console.error(error.message);
+    },
+  });
+  return (
+    <Box bg={colors.background.light} py={7} px={10} minH="100vh">
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:71]: bg="background.light"

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+          <HStack justifyContent="space-between">
+            <Button leftIcon={<ChevronLeftIcon />} variant="link">
+              Back to volunteer postings
+            </Button>
+            <Button>Submit availability</Button>
+          </HStack>
+        </Container>
+
+        <Container
+          bg={colors.background.white}
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:83]: bg="background.white"

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+          borderRadius={10}
+        >
+          <VStack w="full">
+            {postingDetails ? (
+              <Box p={6} w="full">
+                <VStack alignItems="start" w="full">
+                  <VStack alignItems="start" marginBottom={4} w="full">
+                    <HStack justifyContent="space-between" w="full">
+                      <Tag>{postingDetails.branch.name}</Tag>
+                      <Text textStyle="caption" color={colors.text.gray}>
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:95]: color="text.gray"

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                        {skill.name}
+                      </Tag>
+                    ))}
+                  </HStack>
+                  <Text textStyle="body-regular" py={4}>
+                    {postingDetails.description}
+                  </Text>
+                  <Text textStyle="body-regular">Point(s) of contact:</Text>
+                  <HStack pb={4}>
+                    {/* {postingDetails.employees.map((employee) => <PocCard />)} */}
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:129]: remove this comment. we can add this code back when we need it

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                      name="John Doe"
+                      title="hard code"
+                      email="hard code"
+                      phoneNumber="hard code"
+                      key={1}
+                    />
+                  </HStack>
+                  <Divider />
+                  <HStack justifyContent="space-between" pt={4} w="full">
+                    <Text textStyle="caption" color={colors.text.gray}>
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:147]: color="text.gray"

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    setPostingDetails,
+  ] = useState<PostingResponseDTO | null>(null);
+  useQuery(POSTING, {
+    variables: { id },
+    fetchPolicy: "cache-and-network",
+    onCompleted: (data) => {
+      setPostingDetails(data.posting);
+    },
+    onError: (error) => {
+      console.error(error.message);
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:67]: hm not sure if we have a way of logging errors in the frontend. I'll get back to you on this, but we shouldn't be console logging them

File: frontend/src/types/api/BranchType.ts [hunk 8]
```diff
@@ -0,0 +1,8 @@
+export type BranchDTO = {
+  id: string;
+  name: string;
+};
+
+export type BranchRequestDTO = Omit<BranchDTO, "id">;
+
+export type BranchResponseDTO = BranchDTO;
```
Comment (LenaNguyen) [inline frontend/src/types/api/BranchType.ts:8]: Just checking, these types aren't being used currently, right? You're just adding them preemptively? Same with the types for Shifts and Skills

**Review by LenaNguyen — CHANGES_REQUESTED** [review]
Run `yarn fix` after addressing the comments and rebasing to help with the linting warnings :)

File: frontend/src/App.tsx [hunk 9]
```diff
@@ -72,74 +73,69 @@ const App = (): React.ReactElement => {
           <AuthContext.Provider
             value={{ authenticatedUser, setAuthenticatedUser }}
           >
-            <PostingContext.Provider value={postingContext}>
-              <PostingContextDispatcherContext.Provider
```
[the remark this answers is not in the document]
Comment (albertlai431) [reply frontend/src/App.tsx]: Oops must've been a mistake on my end when rebasing, it's fixed now!

File: frontend/src/components/common/PocCard.tsx [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        <PhoneIcon color="text.gray" w="18px" h="18px" />
+        <Text textStyle="caption" fontSize="14px">
+          {phoneNumber}
+        </Text>
+      </HStack>
+    </Stack>
+  );
+};
+
+export default PocCard;
```
[the remark this answers is not in the document]
Comment (albertlai431) [reply frontend/src/components/common/PocCard.tsx]: Yep, just rebased off main so it's updated with his version now

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx (hunk 7 above)
Comment (albertlai431) [reply frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx]: For visibility: I ended up just removing the error handling, to be addressed in a future ticket most likely

File: frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+  Button,
+} from "@chakra-ui/react";
+
+import { gql, useQuery } from "@apollo/client";
+import { useParams } from "react-router-dom";
+import { CalendarIcon, TimeIcon, ChevronLeftIcon } from "@chakra-ui/icons";
+import { PostingResponseDTO } from "../../../../types/api/PostingTypes";
+import PocCard from "../../../common/PocCard";
+
+// Description is the markdown description
```
Comment (LenaNguyen) [inline frontend/src/components/pages/volunteer/posting/VolunteerPostingDetails.tsx:19]: does this comment need to be here? Also isn't the description rich text?

**Review by LenaNguyen — CHANGES_REQUESTED** [review]
Great work! Just a couple small things that I missed in my first review.

**Review by LenaNguyen — APPROVED** [review]
Those were some speedy changes! Great work!

**Pull request merged**